### PR TITLE
feat: Initial support for Scala 3

### PR DESCRIPTION
**Pull request by Johnnei**
Very minimal initial support for Scala 3 (LTS 3.3 and latest 3.4)

- Split source into sharable components (reporting / configuration) and Scala 2/3 specific code
- Set up InspectionTest that runs Scapegoat as true compiler plugin (with extra wrapping to extract the state without having to parse reports)
- Set up base inspection traversal to help with extraction of snippets
- Port OptionGet inspection to Scala 3

Some "obvious" omissions:
- A lot of the Scala 2 generic traversal code is still missing so it might still traverse too aggressively
- Because of that Suppression of  warnings is still unsupported too

I didn't squash the commits to allow you to see my journey and maybe help a bit with reviewing the steps I took. But please squash on merge :)

### CODE REVIEW

File: .github/workflows/pr-checks.yml [hunk 1]
```diff
@@ -14,6 +14,8 @@ jobs:
           - 2.12.19
           - 2.13.13
           - 2.13.14
+          - 3.3.3
+          - 3.4.2
```
Comment (Johnnei) [inline .github/workflows/pr-checks.yml:18]: My github action-voo is not amazing, probably want to only run Scala 3 with 11. If people are modern enough to use Scala 3, then surely they moved on from Java 8.

File: .scalafmt.conf [hunk 2]
```diff
@@ -1,5 +1,5 @@
 version = "3.8.3"
-runner.dialect = scala213source3
+runner.dialect = scala3
```
Comment (Johnnei) [inline .scalafmt.conf:2]: looks a bit weird, but seems to still format Scala 2 sources.

File: build.sbt [hunk 3]
```diff
@@ -22,15 +19,16 @@ developers := List(
   )
 )
 
-scalaVersion := "2.13.14"
```
Comment (Johnnei) [inline build.sbt:25]: not sure if this is too optimistic (especially for codecov?)

**Review by saeltz — COMMENTED** [review]
Thank you very much for your effort! This looks great. I have added some comments.

File: src/main/scala-3/com/sksamuel/scapegoat/inspections/option/OptionGet.scala [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import com.sksamuel.scapegoat.*
+import dotty.tools.dotc.ast.Trees.*
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.StdNames.*
+import dotty.tools.dotc.util.SourcePosition
+
+/**
+ * @author
+ *   Stephen Samuel
```
[the remark this answers is not in the document]
Comment (Johnnei) [reply src/main/scala-3/com/sksamuel/scapegoat/inspections/option/OptionGet.scala:12]: Don't really care to be honest, git blame tends to be more correct than author annotations. But this came from copy paste, I took the scala 2 variant as base to port.

[the remark this answers is not in the document]
Comment (sksamuel) [reply src/main/scala-3/com/sksamuel/scapegoat/inspections/option/OptionGet.scala:12]: Feel free to remove my name

File: build.sbt [hunk 5]
```diff
@@ -65,6 +63,7 @@ scalacOptions := {
   common ++ (scalaBinaryVersion.value match {
     case "2.12" => scalac12Options
     case "2.13" => scalac13Options
+    case "3"    => Seq()
```
[the remark this answers is not in the document]
Comment (Johnnei) [reply build.sbt:66]: Yep, still need to clean up a bit more here and there.

[the remark this answers is not in the document]
Comment (Johnnei) [reply build.sbt]: Added `-Wconf` as well to remove some non-actionable warnings but seem it isn't as flexible to configure as the Scala 2.13 variant yet.  
Still a few warnings that we can't fix as we're cross-compiling to Scala 2.12 (scala 3 syntax migrations in the shared code).

File: src/main/scala-3/com/sksamuel/scapegoat/Plugin.scala [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+  private[scapegoat] var feedback: Option[FeedbackDotty] = None
+
+  override def phaseName: String = "scapegoat"
+
+  override val runsAfter: Set[String] = Set(TyperPhase.name)
+
+  override val runsBefore: Set[String] = Set(PatternMatcher.name)
+
+  override def run(using Context): Unit = {
```
Comment (Johnnei) [inline src/main/scala-3/com/sksamuel/scapegoat/Plugin.scala:44]: todo: Need to double check this flow. Seems to write reports far too often in the CI check, is this running per file?

**Comment by Johnnei** [comment]
@saeltz Could you do another round of review? I think should be minimum viable now to get Scala 3 support going.

Slight decrease in project codecov which I think is mostly the Scala 2.13 version checks which don't trigger in Scala 3.

**Review by saeltz — APPROVED** [review]
Thanks a ton!
Let's get this merged and add more and more inspections for Scala 3 iteratively.

**Comment by saeltz** [comment]
@mccartney, would you issue a new release? Maybe even a minor version? We should also properly announce this.

**Comment by t1b00** [comment]
Hey, really cool to see that you are considering a port to Scala 3. 
As part of my semester project, I actually looked into porting Scapegoat to Scala 3 as this was an issue for our project and for the Scala community. However, I chose to use the Scalafix tool developed by the Scalacenter. 
Would you be interested in collaborating? I've ported 36 rules for now, with functioning tests. Maybe that could shave off some workload for you ;) https://github.com/dedis/scapegoat-scalafix

**Comment by mccartney** [comment]
@Johnnei wow, this is great! What a nice surprise.
@saeltz Sorry, I must've missed the notification.

Will perform the release and will report back. I think it should be a major one, even if there's no specific backward incompatibility we are aware of.

**Comment by mccartney** [comment]
[3.0.0](https://github.com/scapegoat-scala/scapegoat/releases/tag/v3.0.0) is on its way.

**Comment by t1b00** [comment]
> @t1b00 Cool to see this interest! It would be great to have more contributors to get Scala 3 support expanded.
> 
> I do think that using Scalameta trees is a nice step forward, as it should (as far I understand) enable more code re-use between Scala 2 and Scala 3 inspections.
> 
> However, I do think that Scapegoat's reporting is one aspect that wouldn't hold as custom Scalafix rules. One of the hard things with Static Analysis is adoption on existing projects and changing rulesets over time. With Scapegoat's customizable Rule Severity, projects can gradually adopt a more enforcing ruleset without breaking the build.
> 
> This is something that helped me to scale this to hundreds of projects in my company in an "advising" mode, rather than build breaker.
> 
> To me, one thing that I really like about Scalafix is the capability that it can fix its own findings. While for plenty of static analysis rules, there's no generic automatic fix. Only an advise on how to approach the solution. I'd suspect that it would be quite a lessened experience compared to Scalafix' out-of-the-box rules.
> 
> So I'm quite interested to see if incorporating Scalameta is feasible, but not sure relying fully on Scalafix will provide the best experience.
> 
> @mccartney @saeltz Interested to hear your thoughts on this as well.

Hey, thanks for your answer!

Let me clarify a few things about Scalafix and Scalameta since they are a bit opaque and I've acquired quite a bit of experience overtime with them.
> "With Scapegoat's customizable Rule Severity, projects can gradually adopt a more enforcing ruleset without breaking the build."
>
That's also the case in Scalafix! Rules can have a severity of Info, Warning or Error, with only error causing compilation errors. Furthermore, it's trivial to activate / deactive rules as one can simply list them in their Scalafix configuration file.

> One thing that I really like about Scalafix is the capability that it can fix its own findings.
>
It can, if you write rewrite rules for it. Scalafix is not a linter, but a linter building tool and it was designed for the community to build their own rule sets, not with the idea of it becoming a linter :) 
For my linter project, I actually focused on the linting part, not rewriting, because as you are saying:
> While for plenty of static analysis rules, there's no generic automatic fix.
>
Scalafix allows you to write linter rules **or** rewrites, but you can simply stick to linter rules if you feel that there is no standard rewrite. Though I could extend some of the rules (e.g. MapGetAndGetOrElse) to a rewrite.

I'd be really happy to collaborate with the Scapegoat team on this if you feel like porting to Scalafix and Scalameta is in your interest.

**Comment by saeltz** [comment]
> I'd be really happy to collaborate with you and the rest of the Scapegoat team on this if you feel like porting to Scalafix and Scalameta is in your interest.

I'd be happy to see this happening. Anything that brings us closer to full Scala 3 support is desirable.

**Comment by t1b00** [comment]
> I'd be happy to see this happening. Anything that brings us closer to full Scala 3 support is desirable.
Amazing! I think adding my code to the Scapegoat repo would be quite cumbersome as the setup is quite different, but you could definetly indicate this as the WIP version for Scapegoat for Scala 3 if that suits you. Otherwise we can also discuss different modalities.
Feel free to contact me by email: thibault.czarniak@epfl.ch

**Pull request merged**